tests/test_qnn.py

```diff
@@ -0,0 +1,429 @@
+    def test_str_repr(self, get_circuit, output_dim):
+        """Test the __str__ and __repr__ representations"""
+        c, w = get_circuit
+        layer = KerasLayer(c, w, output_dim)
+
+        assert layer.__str__() == "<Quantum Keras layer: func=circuit>"
+        assert layer.__repr__() == "<Quantum Keras layer: func=circuit>"
+
+
+@min_tf
+@pytest.mark.usefixtures("get_circuit", "model")
+class TestKerasLayerIntegration:
+    """Integration tests for the pennylane.qnn.KerasLayer class."""
+
+    @pytest.mark.parametrize("interface", qml.qnodes.decorator.ALLOWED_INTERFACES)
+    @pytest.mark.parametrize("n_qubits, output_dim", indices(2))
+    @pytest.mark.parametrize("batch_size", [5, 10])
+    def test_train_model(self, model, batch_size, n_qubits, output_dim):
+        """Test if a model can train using the KerasLayer. The model is composed of a single
+        KerasLayer sandwiched between two Dense layers, and the dataset is simply input and output
+        vectors of zeros. The test checks that the loss function after two epochs is less than
+        the loss function after one epoch, indicating that training is taking place."""
+
+        x = np.zeros((5, n_qubits))
+        y = np.zeros((5, output_dim))
+
+        model.compile(optimizer="sgd", loss="mse")
+
+        result = model.fit(x, y, epochs=2, batch_size=batch_size, verbose=0)
+        loss = result.history["loss"]
+
+        assert loss[0] > loss[-1]
+
+    @pytest.mark.parametrize("interface", qml.qnodes.decorator.ALLOWED_INTERFACES)
+    @pytest.mark.parametrize("n_qubits, output_dim", indices(2))
+    def test_model_gradients(self, model, output_dim, n_qubits):
+        """Test if a gradient can be calculated with respect to all of the trainable variables in
+        the model"""
+        x = tf.zeros((5, n_qubits))
+        y = tf.zeros((5, output_dim))
+
+        with tf.GradientTape() as tape:
+            out = model(x)
+            loss = tf.keras.losses.mean_squared_error(out, y)
+
+        gradients = tape.gradient(loss, model.trainable_variables)
+        assert all([not isinstance(g, type(None)) for g in gradients])
+
+    @pytest.mark.parametrize("interface", ["tf"])
+    @pytest.mark.parametrize("n_qubits, output_dim", indices(2))
+    def test_model_save_weights(self, model, n_qubits):
+        """Test if the model can be successfully saved and reloaded using the get_weights()
+        method"""
+        _, filename = tempfile.mkstemp()
+        prediction = model.predict(np.ones(n_qubits))
+        weights = model.get_weights()
+        model.save_weights(filename)
+        model.load_weights(filename)
+        prediction_loaded = model.predict(np.ones(n_qubits))
+        weights_loaded = model.get_weights()
+
+        assert np.allclose(prediction, prediction_loaded)
+        for i, w in enumerate(weights):
+            assert np.allclose(w, weights_loaded[i])
+
+        os.remove(filename)
```

Comment:
I moved the `os.remove` above the `assert` so that the removal should not get interrupted.

Although I'm not sure what you mean about using `tmpdir`?

Reply:
Pytest provides a `tmpdir` fixture, that takes care of the set up and tear down for you. That is, if something fails in the test (even before the `assert` line is called), the fixture ensures that all temporary files and directories are still safely deleted.

See https://docs.pytest.org/en/latest/tmpdir.html

Reply:
Awesome! Have used this now.